### ISSUE TITLE
group library errors

### ISSUE THREAD

**Issue by nevrome** [crsh/citr#25]
Thanks for 6730d5a! I really appreciate the fast implementation of this feature. Unfortunately I get errors when I try to load group libraries. Interesting enough there's a different error for each of the three group libraries I'm currently using. Nevertheless I decided to only create one issue here: I guess they are somehow related. 

**Edit:** I just realized that the errors appear interchangeably. They are not database specific.

**Library A: 9061 items**
```
Listening on http://127.0.0.1:6537
Warning: Error in do_read_bib: lex fatal error:
input buffer overflow, can't enlarge buffer because scanner uses REJECT

Stack trace (innermost first):
    105: do_read_bib
    104: RefManageR::ReadBib
    103: import_bbt_piecewise
    102: load_betterbiblatex_bib
    101: eval
    100: eval
     99: shiny::withProgress
     98: <reactive>
     87: bibliography
     86: force
     85: need
     84: lapply
     83: sapply
     82: validate
     81: renderText
     80: func
     79: origRenderFunc
     78: output$read_error
      3: shiny::runApp
      2: runGadget
      1: citr:::insert_citation
Warning: Error in do_read_bib: lex fatal error:
input buffer overflow, can't enlarge buffer because scanner uses REJECT

Stack trace (innermost first):
    61: <Anonymous>
    60: stop
    59: bibliography
    58: observerFunc
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
Warning: Error in do_read_bib: lex fatal error:
input buffer overflow, can't enlarge buffer because scanner uses REJECT

Stack trace (innermost first):
    91: <Anonymous>
    90: stop
    89: bibliography
    88: tag
    87: tags$a
    86: actionLink
    85: tag
    84: tags$span
    83: span
    82: helpText
    81: renderUI
    80: func
    79: origRenderFunc
    78: output$zotero_status
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
```

**Library B: 10967 items**
```
Listening on http://127.0.0.1:6537
Warning: Error in do_read_bib: lex fatal error:
fatal flex scanner internal error--end of buffer missed

Stack trace (innermost first):
    105: do_read_bib
    104: RefManageR::ReadBib
    103: import_bbt_piecewise
    102: load_betterbiblatex_bib
    101: eval
    100: eval
     99: shiny::withProgress
     98: <reactive>
     87: bibliography
     86: force
     85: need
     84: lapply
     83: sapply
     82: validate
     81: renderText
     80: func
     79: origRenderFunc
     78: output$read_error
      3: shiny::runApp
      2: runGadget
      1: citr:::insert_citation
Warning: Error in do_read_bib: lex fatal error:
fatal flex scanner internal error--end of buffer missed

Stack trace (innermost first):
    61: <Anonymous>
    60: stop
    59: bibliography
    58: observerFunc
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
Warning: Error in do_read_bib: lex fatal error:
fatal flex scanner internal error--end of buffer missed

Stack trace (innermost first):
    91: <Anonymous>
    90: stop
    89: bibliography
    88: tag
    87: tags$a
    86: actionLink
    85: tag
    84: tags$span
    83: span
    82: helpText
    81: renderUI
    80: func
    79: origRenderFunc
    78: output$zotero_status
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
```

**Library C: 19 items**
```
Listening on http://127.0.0.1:6537
<simpleError in do_read_bib(file, encoding = .Encoding, srcfile): lex fatal error:
fatal flex scanner internal error--end of buffer missed
>
Warning: Error in if: argument is of length zero
Stack trace (innermost first):
    67: observeEventHandler
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
```

**devtools::session_info()**
```
setting  value                       
 version  R version 3.4.3 (2017-11-30)
 system   x86_64, linux-gnu           
 ui       RStudio (1.1.379)           
 language (EN)                        
 collate  en_US.utf8                  
 tz       Europe/Berlin               
 date     2018-01-09                  

Packages ------------------------------------------------------------------------------------------------------------------
 package    * version    date       source                            
 assertthat   0.2.0      2017-04-11 CRAN (R 3.4.0)                    
 backports    1.1.2      2017-12-13 CRAN (R 3.4.3)                    
 base       * 3.4.3      2017-11-30 local                             
 bibtex       0.4.2      2017-06-30 CRAN (R 3.4.1)                    
 bookdown     0.5.10     2017-12-03 Github (rstudio/bookdown@d78c661) 
 citr         0.2.0.9055 2018-01-06 Github (crsh/citr@6730d5a)        
 compiler     3.4.3      2017-11-30 local                             
 curl         3.1        2017-12-12 cran (@3.1)                       
 datasets   * 3.4.3      2017-11-30 local                             
 devtools     1.13.4     2017-11-09 CRAN (R 3.4.2)                    
 digest       0.6.13     2017-12-14 CRAN (R 3.4.3)                    
 evaluate     0.10.1     2017-06-24 CRAN (R 3.4.0)                    
 graphics   * 3.4.3      2017-11-30 local                             
 grDevices  * 3.4.3      2017-11-30 local                             
 htmltools    0.3.6      2017-04-28 CRAN (R 3.4.0)                    
 httpuv       1.3.5      2017-07-04 CRAN (R 3.4.1)                    
 httr         1.3.1      2017-08-20 cran (@1.3.1)                     
 jsonlite     1.5        2017-06-01 CRAN (R 3.4.0)                    
 knitr        1.18       2017-12-27 CRAN (R 3.4.3)                    
 lubridate    1.7.1      2017-11-03 CRAN (R 3.4.2)                    
 magrittr     1.5        2014-11-22 CRAN (R 3.4.3)                    
 memoise      1.1.0      2017-04-21 CRAN (R 3.4.2)                    
 methods    * 3.4.3      2017-11-30 local                             
 mime         0.5        2016-07-07 CRAN (R 3.4.0)                    
 miniUI       0.1.1      2016-01-15 CRAN (R 3.4.0)                    
 plyr         1.8.4      2016-06-08 CRAN (R 3.4.2)                    
 R6           2.2.2      2017-06-17 CRAN (R 3.4.0)                    
 Rcpp         0.12.14    2017-11-23 cran (@0.12.14)                   
 RefManageR   0.14.20    2017-08-17 CRAN (R 3.4.2)                    
 rmarkdown    1.8.3      2017-12-03 Github (rstudio/rmarkdown@07f7d8e)
 rprojroot    1.3-2      2018-01-03 CRAN (R 3.4.3)                    
 rstudioapi   0.7        2017-09-07 CRAN (R 3.4.2)                    
 shiny      * 1.0.5      2017-08-23 CRAN (R 3.4.2)                    
 stats      * 3.4.3      2017-11-30 local                             
 stringi      1.1.6      2017-11-17 CRAN (R 3.4.2)                    
 stringr      1.2.0      2017-02-18 CRAN (R 3.4.0)                    
 tools        3.4.3      2017-11-30 local                             
 utils      * 3.4.3      2017-11-30 local                             
 withr        2.1.1      2017-12-19 CRAN (R 3.4.3)                    
 xml2         1.1.1      2017-01-24 CRAN (R 3.4.0)                    
 xtable       1.8-2      2016-02-05 CRAN (R 3.4.1)                    
 yaml         2.1.16     2017-12-12 cran (@2.1.16)
```

**Comment by crsh** [comment]
Hi Clemens, thanks for the feedback! I assume this is related to long bib-entries. The underlying functions I use to read the BibTeX files run into problems with those sometimes. I've been trying to get around this, but it seems it failed in your case. Could you do me a favor and send me a copy of the files by accessing them in your browser? For example, try `http://localhost:23119/better-bibtex/library?/9061/library.bibtex`.

**Comment by nevrome** [comment]
Thanks Frederik. I wrote you an email.

**Comment by nevrome** [comment]
Is there a way to locate the entries in my database that cause problems?

**Comment by crsh** [comment]
Hi Clemens, sorry I've been swamped. I would try the divide-and-conquer-strategy. Cut the file in half and see if you can read it. If you cannot, cut the remainder in half and so on.

**Comment by crsh** [comment]
Let me know if you can locate the problem.

**Comment by nevrome** [comment]
OK - that's a strategy... I guess I could maybe even automate this somehow... Hm.

**Comment by nevrome** [comment]
There are several citations in my DB with very long entries in the abstract field. Maybe that caused some of the problems. I decided now to create a small, clean subset of the literature that I need for the task at hand. That works flawlessly with citr, is faster and I don't have to clean my complete database. 

Altogether zotero gave me a pretty hard time in the last couple of days. I'm disillusioned with group libraries.

Thank you for the help @crsh! The issue can be closed although I would not consider this problem solved.

**Comment by crsh** [comment]
Ah, yes, I've had this kind of issue come up before. This is a limitation of the R package I use (the only one I know) to read the Bib(La)TeX files. With Better Bib(La)TeX you can set an option to not export the abstract field, which should fix this problem once and for all without requiring manual labor.

![bbt_omit_abstract](https://user-images.githubusercontent.com/2195485/35349341-ecee16ac-013a-11e8-8832-f5c01f40c5ef.png)

I should probably document this somewhere...

**Comment by nevrome** [comment]
Cool - good to know. I get a new error message now: 

```
Listening on http://127.0.0.1:5052
Warning in do_read_bib(file, encoding = .Encoding, srcfile) :
  
bASQEt1ZQW5hQEgcTxGlbG7pzYGk9eSr.bib:2:0
	syntax error, unexpected $end
	Dropping the entry `(null)` (starting at line 0) 
Warning: Error in do_read_bib: unprotect_ptr: pointer not found
Stack trace (innermost first):
    107: do_read_bib
    106: RefManageR::ReadBib
    105: import_bbt_piecewise
    104: load_betterbiblatex_bib
    103: eval
    102: eval
    101: shiny::withProgress
    100: <reactive>
     89: bibliography
     88: tag
     87: tags$a
     86: actionLink
     85: tag
     84: tags$span
     83: span
     82: helpText
     81: renderUI
     80: func
     79: origRenderFunc
     78: output$zotero_status
      3: shiny::runApp
      2: runGadget
      1: citr:::insert_citation
Warning: Error in do_read_bib: unprotect_ptr: pointer not found
Stack trace (innermost first):
    89: <Anonymous>
    88: stop
    87: bibliography
    86: force
    85: need
    84: lapply
    83: sapply
    82: validate
    81: renderText
    80: func
    79: origRenderFunc
    78: output$read_error
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
Warning: Error in do_read_bib: unprotect_ptr: pointer not found
Stack trace (innermost first):
    61: <Anonymous>
    60: stop
    59: bibliography
    58: observerFunc
     3: shiny::runApp
     2: runGadget
     1: citr:::insert_citation
```

bASQEt1ZQW5hQEgcTxGlbG7pzYGk9eSr.bib is created and not deleted. It contains only the line
```
translation failed
```

**Comment by crsh** [comment]
Not sure what's going on and I can't really address the issue unless I can reproduce it. Could you upload a copy of the bibfile that produces the error? You can access the file at `http://localhost:23119/better-bibtex/library?/` + the library ID + `/library.bibtex`.

**Comment by nevrome** [comment]
I already sent you this via email as far as I remember.

**Comment by crsh** [comment]
But this error now occured after changing the export settings, no? So it would yield a different BibTeX file and facilitate pinpointing the problem.

**Comment by nevrome** [comment]
You're right. I understand. Sent you another mail. Thanks!

**Comment by crsh** [comment]
Hi Clemens, I finally found some time to look into this again. The new file you sent me (without the abstracts) works for me, whereas the old file does not.

The problem with very long fields in .bib-files is a [known issue](https://github.com/romainfrancois/bibtex/issues/16) of `bibtex::read.bib()` that `citr` builds on. I don't think there is anything I can do about it. I have, however, included a note on how to fix this problem to the README.

Concerning the new file: Could you do me a favor and try one more time? I have [noticed](https://github.com/crsh/citr#known-problems) that once `read.bib()` fails with a lex-error it is sometimes necessary to restart the R session. Otherwise reading of any .bib-file fails. It's the first thing I can think of, why this would work on my machine but not yours.

**Comment by nevrome** [comment]
I tried and first of all it failed again with the not very speaking error message mentioned above: 
`Warning: Error in do_read_bib: unprotect_ptr: pointer not found`

Then I selected the library of interest in the library tree panel of Zotero. That did the trick. It's working now. GUI magic. I guess selecting it triggers BetterBibtex to provide the bibfile. Something pretty harmless and still annoying. Maybe you even mention this somewhere in your description and I just forgot the obvious. 

Thanks a lot for your help!
